### DISCUSSION TITLE
Ideas for v2.x

### DISCUSSION THREAD

**Discussion by VonHeikemen** [VonHeikemen/lsp-zero.nvim#130] [General]
For the next version of `lsp-zero` I will to move away from this opionionated model v1 has. After reading a little bit of the source code of `nvim-lspconfig` and `nvim-cmp` I realize I can "extend" them instead of controlling them.

I would like `lsp-zero` to be a collection of function that help reduce the boilerplate code.

## LSP setup

Here I would like to keep `.extend_lspconfig()` and `.setup_servers()`. These two will be enough to make the configuration of LSP easy. And when you need to, you can use `lspconfig` directly.

```lua
local lsp_zero = require('lsp-zero')
local lspconfig = require('lspconfig')

lsp_zero.extend_lspconfig({
  set_lsp_keymaps = true,
  on_attach = function(client, bufnr)
    -- You can do stuff here...
  end
})

lspconfig.sumneko_lua.setup(lsp_zero.nvim_workspace())

lsp_zero.setup_servers({'tsserver', 'eslint'})
```

### What about mason.nvim?

I think mason is pretty easy to use.

Consider this code.

```lua
require('mason').setup()
require('mason-lspconfig').setup({automatic_installation = true})

local lsp_zero = require('lsp-zero')
lsp_zero.extend_lspconfig({set_lsp_keymaps = true})

lsp_zero.setup_servers({'tsserver', 'eslint'})
```

Here we have automatic install and configuration of LSP servers. In many cases this would be easy enough. Best thing is, `lsp-zero` is not doing anything special, `.setup_servers()` is literally a `for` loop that calls lspconfig. I know some people struggle with for loops, that's why I want to keep `.setup_servers()`.

Things get a little bit more complicated if the user wants to use the option `ensure_installed` of `mason-lspconfig`. And here I'm wondering: should I add a special syntax to support this case? Something like this.

```lua
require('mason-lspconfig').setup({
  ensure_installed = {'tsserver', 'eslint', 'rust_analyzer'}
})

---
-- some code...
---

lsp_zero.setup_servers({
  installed_with = 'mason',
  omit = {'rust_analyzer'}
})
```

I don't like it... but there's people out there who really like this `ensure_installed` thing.

If you have any ideas for this case, I'll hear you.

## Autocompletion

Turns out nvim-cmp can be extended too. Kind of like lspconfig. This is because `cmp.setup` only modifies a configuration table, it doesn't look like it has side effects. It's safe to call it twice. I can set some defaults and let the user handle `sources` and `mapping`.

I know people are not afraid of copy/paste a piece of code, as long as it works without modification. This could work just fine.

```lua
require('lsp-zero').extend_cmp()

vim.opt.completeopt = {'menu', 'menuone', 'noselect'}
require('luasnip.loaders.from_vscode').lazy_load()

local cmp = require('cmp')
local cmp_action = require('lsp-zero').cmp_action

cmp.setup({
  sources = {
    {name = 'nvim_lsp'},
    {name = 'luasnip'},
  },
  mapping = cmp.mapping.preset.insert({
    ['<C-Space>'] = cmp.mapping.complete(),
    ['<C-u>'] = cmp.mapping.scroll_docs(-4),
    ['<C-d>'] = cmp.mapping.scroll_docs(4),
    ['<C-f>'] = cmp_action.luasnip_jump_forward(),
    ['<C-b>'] = cmp_action.luasnip_jump_backward(),
  })
})
```

It's a lot more verbose than the `recommended` preset, but if it works people won't complain. And this can exists in a different file. It doesn't have any effect on lspconfig. No need to setup LSP and autocompletion in the same place.

And notice the variable `cmp_action`, that one comes from `lsp-zero`. There I can have things like "supertab" and the user would just need to call it.

```lua
['<Tab>'] = cmp_action.luasnip_supertab(),
['<S-Tab>'] = cmp_action.luasnip_shift_supertab(),
```

What do you all think?

## What about the presets?

I'm not against them, I just think that with the right example code in a "getting started" section would be enough to make people happy.

If I do keep them I would like an API that's just a function. Something like this.

```lua
require('lsp-zero').use_preset('recommended', {
  ---
  -- Here we can have some options.
  -- What would you like?
  ---
})
```

**Comment by VonHeikemen** [comment]
Ok. So maybe I can keep one preset, the one that deals with `lspconfig` and `mason`. I kinda like how this looks.

```lua
local lsp = require('lsp-zero').preset({'mason-lsp'})

lsp.on_attach(function(client, bufnr)
  lsp.set_keymaps(bufnr) -- delete this if you want to disable lsp-zero's keymaps
end)

lsp.ensure_installed({
  -- Replace these with whatever servers you want to install
  'tsserver',
  'eslint',
  'rust_analyzer',
  'sumneko_lua',
})

lsp.setup_servers({
  mason_lsp = true,                        -- Configure all servers installed with mason.nvim
  omit = {'sumneko_lua', 'rust_analyzer'}  -- except these.
})

-- If you want to pass arguments to an LSP server use `lspconfig`
require('lspconfig').rust_analyzer.setup({})

-- Configure lua language server for neovim
require('lspconfig').sumneko_lua.setup(lsp.nvim_workspace())
``` 

I think this is  "beginner friendly" 'cause there are less moving parts. Users will only have to deal with the api of 2 plugins instead of 4. And the only thing they need to remember is to call `lspconfig` after `lsp-zero`.

I also think keeping this will make the transition from `v1` to `v2` easier, at least for the LSP setup. `nvim-cmp` will not be included in the preset. People really like to customize nvim-cmp, don't think it would be wise to assume control of that plugin. 

If you want to control the mason plugins, you could opt-out.

```lua
local lsp = require('lsp-zero').preset({
  'mason-lsp',
  mason_setup = false,
  mason_lsp_setup = false,
})
```

But at that point might as well just use the `extend_lspconfig()` function.

**Comment by VonHeikemen** [comment]
Small update on the `.setup_servers()` business:

I'm liking this one.

```lua
local lsp = require('lsp-zero').preset({'mason-lsp'})

lsp.on_attach(function(client, bufnr)
  lsp.default_keymaps(bufnr)
end)

lsp.ensure_installed({
  'tsserver',
  'eslint',
  lsp.skip_setup({
    'sumneko_lua',
    'rust_analyzer',
  })
})

lsp.setup_servers()

require('lspconfig').sumneko_lua.setup(lsp.nvim_workspace())

require('lspconfig').rust_analyzer.setup({})
```

**Comment by VonHeikemen** [comment]
I just realized I could make this work.

```lua
local lsp = {}

lsp.default_keymaps = true

lsp.installer = {
  name = 'mason',
  automatic_setup = true,
  automatic_install = true,
}

lsp.servers = {
  'tsserver',
  'eslint',
  skip_setup = {
    'sumneko_lua',
    'rust_analyzer',
  },
}

lsp.on_attach = function(client, bufnr)
  -- Create your keymaps here...
end

require('lsp-zero').setup(lsp)
```

**Comment by VonHeikemen** [comment]
The setup function is now growing on me. But I don't want to call it "setup" because it's optional. Don't want people to think is mandatory to use it.

So I like this snippet as "getting started" example.

```lua
local lsp_zero = require('lsp-zero')

local lsp_attach = function(client, bufnr)
  lsp_zero.default_keymaps(bufnr)
end

lsp_zero.preset('mason-lsp', {
  automatic_install = true,
  on_attach = lsp_attach,
})

lsp_zero.setup_servers({'tsserver', 'eslint'})

require('lspconfig').sumneko_lua.setup(lsp_zero.nvim_workspace())
```

**Comment by VonHeikemen** [comment]
Per project settings. Not sure how many use them but I think they are cool. They will not be attached to a preset in `v2`.

The idea here is that you "store" your configuration in your `init.lua` .

```lua
-- init.lua

require('lsp-zero').store_config('sumneko_lua', {
  on_init = function(client)
    -- disable formatting. (because of reasons)
    client.server_capabilities.documentFormattingProvider = false
    client.server_capabilities.documentFormattingRangeProvider = false
  end
})
```

Then in your local config you call the function `use`. The configuration in `init.lua` will be merged with the one in your project.

```lua
--- an example local config for neovim

local lsp_zero = require('lsp-zero')
lsp_zero.use('sumneko_lua', lsp_zero.nvim_workspace())
```

(`lsp-zero` doesn't execute files, it only provides functions. The execution of the local config should be handled by another plugin).

**Reply by neilgilmour** [reply]
Interesting idea. Personally I think I'd rather have sane defaults in my main config, and then opt _out_ in the project if I want to, rather than having to opt _in_.

For example I'm often just spinning something up to try it out or make a quick tweak. I don't want to have to put any extra effort in to make my config work. Of course I could have misunderstood your idea here...

**Comment by VonHeikemen** [comment]
And just because, here is [ThePrimeagen's lsp config](https://github.com/ThePrimeagen/init.lua/blob/fceb9c822478cc7b3fc7c8692ccae717dd678047/after/plugin/lsp.lua) but using `v2` (with `automatic_install` instead of `ensure_installed`).

LSP servers will be installed based on use. So tsserver, rust_analyzer and sumneko_lua will be installed if they are missing.

```lua
local lspconfig = require('lspconfig')
local lsp_zero = require('lsp-zero').preset({
  name = 'mason-lsp',
  automatic_install = true,
})

lsp_zero.on_attach(function(client, bufnr)
  local opts = {buffer = bufnr, remap = false}

  vim.keymap.set("n", "gd", function() vim.lsp.buf.definition() end, opts)
  vim.keymap.set("n", "K", function() vim.lsp.buf.hover() end, opts)
  vim.keymap.set("n", "<leader>vws", function() vim.lsp.buf.workspace_symbol() end, opts)
  vim.keymap.set("n", "<leader>vd", function() vim.diagnostic.open_float() end, opts)
  vim.keymap.set("n", "[d", function() vim.diagnostic.goto_next() end, opts)
  vim.keymap.set("n", "]d", function() vim.diagnostic.goto_prev() end, opts)
  vim.keymap.set("n", "<leader>vca", function() vim.lsp.buf.code_action() end, opts)
  vim.keymap.set("n", "<leader>vrr", function() vim.lsp.buf.references() end, opts)
  vim.keymap.set("n", "<leader>vrn", function() vim.lsp.buf.rename() end, opts)
  vim.keymap.set("i", "<C-h>", function() vim.lsp.buf.signature_help() end, opts)
end)

lsp_zero.setup_servers({
  'tsserver',
  'rust_analyzer',
})

lspconfig.sumneko_lua.setup({
  settings = {
    Lua = {
      diagnostics = {
        globals = {'vim'}
      }
    }
  }
})

require('lsp-zero').extend_cmp()

local cmp = require('cmp')

cmp.setup({
  sources = {
    {name = 'nvim_lsp'},
    {name = 'buffer'},
  },
  mapping = cmp.mapping.preset.insert({
    ['<C-Space>'] = cmp.mapping.complete(),
    ['<C-b>'] = cmp.mapping.scroll_docs(-4),
    ['<C-f>'] = cmp.mapping.scroll_docs(4),
  })
})
```

**Reply by VonHeikemen** [reply]
Using `ensure_installed`.

Here you tell which servers you want installed, but you need to specify you want "automatic setup" and then exclude the servers you want to configure yourself.

```lua
local lsp_zero = require('lsp-zero').preset({
  name = 'mason-lsp',
  setup_new_servers = true,
  ignore_setup = {'sumneko_lua'},
})

lsp_zero.ensure_installed({
  'tsserver',
  'rust_analyzer',
  'sumneko_lua',
})

lspconfig.sumneko_lua.setup({
  settings = {
    Lua = {
      diagnostics = {
        globals = {'vim'}
      }
    }
  }
})
```

**Comment by briandipalma** [comment]
Sorry, I've wanted to reply for a week now, it's just been busy at work.

> It's a lot more verbose than the recommended preset

This, to me, is backwards step. One of the biggest advantages of using `lsp-zero` is that I got to delete lots of Lua config.
Honestly my dream is to have just

```lua
local lsp = require('lsp-zero')

lsp.preset('recommended')
lsp.setup()
```

And for *all* the features I could possibly want to be enabled and working.
The only code/config I might have would be to disable something (can't think what TBH) or to change the default configuration/keybindings.

I don't want to have to copy a bunch of code in from a README to get functionality to work.
It's especially odd if you consider a good 90%+ of people will have to copy the same code into their configs.

I agree it helps people understand how everything works but sometimes people are too busy and they just need stuff to work.
We don't all need to be experts at everything. I have no config for Mason for example and yet it all works great and provides huge value. I feel that's the approach `lsp-zero` should take.

In fact I want to raise more questions/suggestions/PRs if you accept my suggestions to increase `lsp-zero`'s LSP/autocomplete default abilities.

**Reply by VonHeikemen** [reply]
> I don't want to have to copy a bunch of code in from a README to get functionality to work.

You won't have to do that. The readme/documentation will still have everything you need to get the features the `recommended` preset supports.

And by the way, `v2` will have minimun breaking changes. The public api will remain intact. I'll mostly drop features that don't have consistent behavior.

> In fact I want to raise more questions/suggestions/PRs if you accept my suggestions to increase lsp-zero's LSP/autocomplete default abilities.

I can hear your suggestions. But I prefer to decrease the level of "control" lsp-zero has over the other plugins.

**Reply by neilgilmour** [reply]
I sit somewhere in between @VonHeikemen and @briandipalma here - I'd like a really simple way to set up sane defaults, and be able to opt out/override where appropriate, however I don't necessarily want it to automatically set up everything I might ever need. A bit like Folke has done with the LazyVim starter, TJ has done with kickstart, and Prime did with his init.lua.

Here's some pseudo-code that gives my ideal scenario;

```lua
local zero = require('lsp-zero')

zero.setup()
```
This would set up;
- the 'recommended' preset,
- keymaps (with descriptions), 
- anything installed through Mason, 
- suggest things to install on entering new buffers,
- connect the relevant formatters, code actions, diagnostics and hovers,
- cmp sources.

And to override things;

```lua
local zero = require('lsp-zero')

zero.overrides = {
    preset = 'another-preset',
    keymap = {
        // add my own keymap here
    },
    formatting = {
        prettierd.with({
            extra_filetypes = { 'php' }
        }),
    },
    cmp = {
            engine = 'luasnip',
            sources = {
                {
                      name = 'path',
                          option = {
                               trailing_slash = true,
                          },
                },
    },
}

zero.setup()
```

**Reply by VonHeikemen** [reply]
What do you think about this?

```lua
local lsp = require('lsp-zero').preset({
  name = 'minimal',
  set_lsp_keymaps = true,
  manage_nvim_cmp = true,
  suggest_lsp_servers = false,
})

lsp.setup()
```

**Reply by briandipalma** [reply]
I moved to LazyVim about 3 weeks ago, I threw away all my config and tried to use it vanilla. Overall I'm quite happy with the experience and have been adding my own config. I miss how lsp-zero would suggest LSP servers when you opened files but I've been able to delete so much lua config that I'm sticking with it. So I won't be raising any suggestions/PRs for lsp-zero unless LazyVim ends up being abandoned. I think I was going to suggest using the telescope LSP functions in the default keymaps if the telescope plugin was installed and adding support for cmp autocomplete in the command line...

**Comment by VonHeikemen** [comment]
New update on version 2. I can keep the config for nvim-cmp inside lsp-zero. The catch? It'll be minimal by default. The keybindings will just emulate Neovim's default. It will only include `cmp-nvim-lsp` as a source. And the documentation window will be disabled (there will be no keybindings to scroll so might as well not have it).

So the "getting started" example will look like this.

```lua
local lsp = require('lsp-zero').preset({name = 'minimal'})

lsp.on_attach(function(client, bufnr)
  lsp.default_keymaps({buffer = bufnr})
end)

lsp.setup()
```

Yes. I want to push the `.on_attach` function right away, so people know they can create the keybindings there. New users can keep the `.default_keymaps()` function, if they want the defaults. Experienced users can delete it and add their own keymaps.

Custom configurations for nvim-cmp can be done using the `cmp` module. And it can be done before or after lsp-zero's setup.
